### PR TITLE
FIX | Removing BinaryFormatter from NetFx

### PR DESCRIPTION
**Pull request by JRahnama**
[BinaryFormatter](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter?view=net-5.0#remarks) is not secure. During testing for net5 and some other tests we noticed it has only been used in `.NETFramework` inside `SqlDependency` and there was no track of it inside NetCore.

According to [MS Documentation](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide#preferred-alternatives) we had couple of alternative options. [XMLSerializer](https://docs.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer?view=net-5.0) can only work on public methods.

[DataContractSerializer ](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer?view=net-5.0) is chosen as it can work on private and public members.

### CODE REVIEW

File: tools/props/Versions.props [hunk 1]
```diff
@@ -52,10 +52,10 @@
   </PropertyGroup>
   <!-- Test Project Dependencies -->
   <PropertyGroup>
-    <MicrosoftDotNetPlatformAbstractionsVersion>3.1.1</MicrosoftDotNetPlatformAbstractionsVersion>
+    <MicrosoftDotNetPlatformAbstractionsVersion>3.1.6</MicrosoftDotNetPlatformAbstractionsVersion>
     <MicrosoftIdentityModelClientsActiveDirectoryVersion>5.2.6</MicrosoftIdentityModelClientsActiveDirectoryVersion>
-    <MicrosoftNETTestSdkVersion>15.9.0</MicrosoftNETTestSdkVersion>
-    <MicrosoftWindowsCompatibilityVersion>3.1.0</MicrosoftWindowsCompatibilityVersion>
+    <MicrosoftNETTestSdkVersion>16.8.3</MicrosoftNETTestSdkVersion>
+    <MicrosoftWindowsCompatibilityVersion>3.1.1</MicrosoftWindowsCompatibilityVersion>
```
Comment (cheenamalhotra) [inline tools/props/Versions.props:58]: Let's keep these changes out of this PR, u can have another PR instead.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlExceptionTest.cs [hunk 2]
```diff
@@ -33,9 +34,9 @@ public void SerializationTest()
             Assert.Equal(e.StackTrace, sqlEx.StackTrace);
         }
 
-
         [Fact]
-        [ActiveIssue("12161", TestPlatforms.AnyUnix)]
+        [ActiveIssue("12161", TargetFrameworkMonikers.Netcoreapp)]
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlExceptionTest.cs:38]: I see there's another PR to remove this test, please revert these changes from this PR to concentrate of NetFx driver changes for BinaryFormatter only.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 3]
```diff
@@ -241,29 +241,33 @@ private static void InvokeCallback(object eventContextPair)
...
+        //-----------------------------------------------
+        // Private Class to add ObjRef as DataContract
+        //-----------------------------------------------
+        [SecurityPermission(SecurityAction.Assert, Flags = SecurityPermissionFlag.RemotingConfiguration)]
+        [DataContract]
+        private class SqlClientObjRef : IExtensibleDataObject
         {
-            public override Type BindToType(string assemblyName, string typeName)
+            [DataMember]
+            private static ObjRef SqlObjRef;
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:253]: NIT: Rename to `s_sqlObjRef`.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 4]
```diff
@@ -241,29 +241,33 @@ private static void InvokeCallback(object eventContextPair)
...
 
-        // ----------------------------------------
-        // Private class for restricting allowed types from deserialization.
-        // ----------------------------------------
 
-        private class SqlDependencyProcessDispatcherSerializationBinder : SerializationBinder
+        //-----------------------------------------------
+        // Private Class to add ObjRef as DataContract
+        //-----------------------------------------------
+        [SecurityPermission(SecurityAction.Assert, Flags = SecurityPermissionFlag.RemotingConfiguration)]
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:248]: Any reason we add this to SqlClientObjRef class?

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 5]
```diff
@@ -241,29 +241,33 @@ private static void InvokeCallback(object eventContextPair)
...
-                else
-                {
-                    throw new ArgumentException("Unexpected type", nameof(typeName));
-                }
+                SqlObjRef = RemotingServices.Marshal(dispatcher);
+            }
+
+            private ExtensionDataObject _extensionData_Value;
+
+            public ExtensionDataObject ExtensionData
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:262]: Any purpose we keep this property? I don't see this being used.

Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:262]: that comes with IExtensibleDataObject. That interface implements that part. I have changed it to use Auto property I believe. 
you can read more [here ](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.iextensibledataobject?view=net-5.0#remarks). I will test to see if  we can use the class without the interface though.

Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs]: Since this is a private class and has an internal usage, if we had more data members in the future, we could decide about the versioning solution later. So, I believe you can remove that interface.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlExceptionTest.cs [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 using System;
 using System.IO;
+using System.Runtime.Serialization;
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlExceptionTest.cs:7]: unnecessary

Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlExceptionTest.cs:7]: thanks I thought it was removed in previous push. I will remove it right away.

**Review by DavoudEshtehari — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 7]
```diff
@@ -241,29 +240,39 @@ private static void InvokeCallback(object eventContextPair)
...
-                }
+                s_sqlObjRef = RemotingServices.Marshal(dispatcher);
+                _typeInfo = s_sqlObjRef.TypeInfo;
+            }
+
+            internal static bool CanCastToSqlDependencyProcessDispatcher()
+            {
+                return _typeInfo.CanCastTo(typeof(SqlDependencyProcessDispatcher), s_sqlObjRef);
             }
+
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:266]: We can make the following a property, right?

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 8]
```diff
@@ -597,11 +605,13 @@ private static void ObtainProcessDispatcher()
...
                             // Serialize and set in native.
-                            ObjRef objRef = GetObjRef(_processDispatcher);
-                            BinaryFormatter formatter = new BinaryFormatter();
-                            MemoryStream stream = new MemoryStream();
-                            GetSerializedObject(objRef, formatter, stream);
-                            SNINativeMethodWrapper.SetData(stream.GetBuffer()); // Native will be forced to synchronize and not overwrite.
+                            using (MemoryStream stream = new MemoryStream())
+                            {
+                                SqlClientObjRef objRef = new SqlClientObjRef(_processDispatcher);
+                                DataContractSerializer formatter = new DataContractSerializer(objRef.GetType());
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:611]: Let's rename `formatter` where it's replaced by a `DataContractSerializer`.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 9]
```diff
@@ -628,37 +638,41 @@ private static void ObtainProcessDispatcher()
...
+                {
+                    DataContractSerializer formatter = new DataContractSerializer(typeof(SqlClientObjRef));
+                    if (SqlClientObjRef.CanCastToSqlDependencyProcessDispatcher())
+                    {
+                        // Deserialize and set for appdomain.
+                        _processDispatcher = GetDeserializedObject(formatter, stream);
+                    }
+                    else
+                    {
+                        throw new ArgumentException("Unexpected type", nameof(SqlClientObjRef._typeInfo));
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:651]: This exception string should be in the resource file (my mistake :blush:).

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 10]
```diff
@@ -628,37 +638,41 @@ private static void ObtainProcessDispatcher()
...
-        private static SqlDependencyProcessDispatcher GetDeserializedObject(BinaryFormatter formatter, MemoryStream stream)
+        private static SqlDependencyProcessDispatcher GetDeserializedObject(DataContractSerializer formatter, MemoryStream stream)
         {
-            // Use a custom SerializationBinder to restrict deserialized types to SqlDependencyProcessDispatcher.
-            formatter.Binder = new SqlDependencyProcessDispatcherSerializationBinder();
-            object result = formatter.Deserialize(stream);
+            object refResult = formatter.ReadObject(stream);
+            var result = RemotingServices.Unmarshal((refResult as SqlClientObjRef).GetObjRef());
+
             Debug.Assert(result.GetType() == typeof(SqlDependencyProcessDispatcher), "Unexpected type stored in native!");
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:674]: Seems like this assert is not needed anymore, as it'll only be reached if `SqlClientObjRef.CanCastToSqlDependencyProcessDispatcher()` is true, and so the assert will always pass.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs [hunk 11]
```diff
@@ -241,29 +240,39 @@ private static void InvokeCallback(object eventContextPair)
...
+                s_sqlObjRef = RemotingServices.Marshal(dispatcher);
+                _typeInfo = s_sqlObjRef.TypeInfo;
+            }
+
+            internal static bool CanCastToSqlDependencyProcessDispatcher()
+            {
+                return _typeInfo.CanCastTo(typeof(SqlDependencyProcessDispatcher), s_sqlObjRef);
             }
+
+            internal ObjRef GetObjRef()
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:267]: We can make this a property, right?

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs (hunk 8 above)
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDependency.cs:611]: usually formatter comes from IFormatter. I did not want to change that, but if you insist I can rename it to serializer.

**Review by johnnypham — APPROVED** [review]

**Review by karinazhou — APPROVED** [review]
